src/Temporalio/Client/WorkerTaskReachability.cs

```diff
@@ -0,0 +1,133 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+
+namespace Temporalio.Client
+{
+    /// <summary>
+    /// Contains information about the reachability of some Build IDs.
+    /// </summary>
+    /// <param name="BuildIdReachability">Maps Build IDs to information about their reachability.</param>
+    public record WorkerTaskReachability(IReadOnlyDictionary<string, BuildIdReachability> BuildIdReachability)
+    {
+        /// <summary>
+        /// Convert from proto.
+        /// </summary>
+        /// <param name="proto">Proto.</param>
+        /// <returns>Converted value.</returns>
+        internal static WorkerTaskReachability FromProto(Api.WorkflowService.V1.GetWorkerTaskReachabilityResponse proto)
+        {
+            var idMap = proto.BuildIdReachability.ToDictionary(
+                bid => bid.BuildId,
+                bid =>
+                {
+                    var unretrieved = new List<string>();
+                    var tqrDict = bid.TaskQueueReachability
+                        .SkipWhile(tqr =>
+                        {
+                            if (tqr.Reachability.Count == 1 &&
+                                tqr.Reachability[0] == Api.Enums.V1.TaskReachability.Unspecified)
+                            {
+                                unretrieved.Add(tqr.TaskQueue);
+                                return true;
+                            }
+
+                            return false;
+                        })
+                        .ToDictionary(
+                            tqr => tqr.TaskQueue,
+                            tqr =>
+                                tqr.Reachability.Select(r => TaskReachabilityType.FromProto(r)).ToList().AsReadOnly());
+                    return new BuildIdReachability(tqrDict, unretrieved.AsReadOnly());
+                });
+            return new(idMap);
+        }
+    }
+
+    /// <summary>
+    /// Contains information about the reachability of a specific Build ID.
+    /// </summary>
+    /// <param name="TaskQueueReachability">Maps Task Queue names to the reachability status of the Build ID on that
+    /// queue. If the value is an empty list, the Build ID is not reachable on that queue.</param>
+    /// <param name="UnretrievedTaskQueues">If any Task Queues could not be retrieved because the server limits the
+    /// number that can be queried at once, they will be listed here.</param>
+    public record BuildIdReachability(
+        IReadOnlyDictionary<string, ReadOnlyCollection<TaskReachabilityType>> TaskQueueReachability,
+        IReadOnlyCollection<string> UnretrievedTaskQueues);
+
+    /// <summary>
+    /// A supertype for different ways a task might reach Workflows.
+    /// </summary>
+    public abstract record TaskReachabilityType
```

Comment:
Can we make this a C# enum? Note in .NET unlike many of our other SDKs, I simple reuse the proto enums because they are actually quite clean in many of these cases (unlike other langauges where they are ugly). But I don't mind if we translate like this either though.

Reply:
I really like this b/c these can be exhaustively pattern matched unlike the enums

Reply:
Such is life of a C# dev using enums, but I think we should still use enums since we're in C# world. We also use C# enums to represent Temporal enums elsewhere in the SDK (and in general it makes sense as the language feature is there).

The reason C# doesn't consider an enum exhaustive is due to internal compiler things I guess which allow it to technically be casted from any int. But I think we can enable named enum exhaustion check by enabling CS8509 and disabling CS8524. But I may be wrong.

I am a bit surprised this sum type exhaustiveness check works. I guess it's some analyzer that just knows this class can't be implemented externally due to the private constructor.